### PR TITLE
CASSANDRASC-114 Make hash algorithm implementation pluggable

### PR DESCRIPTION
**Pull request by yifan-c**
patch by Yifan Cai; reviewed by Francisco Guerrero for CASSANDRASC-114

### CODE REVIEW

File: src/test/java/org/apache/cassandra/sidecar/utils/XXHash32DigestVerifierTest.java [hunk 1]
```diff
@@ -52,40 +51,37 @@ class XXHash32DigestVerifierTest
...
     static void setup() throws IOException
     {
         vertx = Vertx.vertx();
 
-        randomFilePath = TestFileUtils.prepareTestFile(tempDir, "random-file.txt", 1024);
-    }
+        restoreJobUtil = new RestoreJobUtil(new Lz4XXHash32Provider());
 
-    @Test
-    void failsWhenDigestIsNull()
```
Comment (yifan-c) [inline src/test/java/org/apache/cassandra/sidecar/utils/XXHash32DigestVerifierTest.java:65]: This test is removed, as the test case is imaginary. In prod code, the verifiers are created by the static create method, which guarantees the digest is non-null.

File: src/main/java/org/apache/cassandra/sidecar/restore/RestoreJobUtil.java [hunk 2]
```diff
@@ -34,23 +34,38 @@
...
     // it is part of upload id and get validated by
     // org.apache.cassandra.sidecar.utils.SSTableUploadsPathBuilder.UPLOAD_ID_PATTERN
     private static final String RESTORE_JOB_PREFIX = "c0ffee-";
     private static final int RESTORE_JOB_PREFIX_LEN = RESTORE_JOB_PREFIX.length();
+    private static final int RESTORE_JOB_DEFAULT_HASH_SEED = 0;
+
+    private DigestAlgorithmProvider digestAlgorithmProvider;
+
+    @Inject
+    public RestoreJobUtil(@Named("xxhash32") DigestAlgorithmProvider digestAlgorithmProvider)
```
Comment (frankgh) [inline src/main/java/org/apache/cassandra/sidecar/restore/RestoreJobUtil.java:65]: something to think for a future improvement. I think the manifest file should have metadata including the hashing algorithm, and we should pick the correct implementation based on that metadata file

File: src/main/java/org/apache/cassandra/sidecar/restore/RestoreJobUtil.java [hunk 3]
```diff
@@ -211,4 +167,60 @@ public static void cleanDirectory(Path path) throws IOException
...
+    /**
+     * @param file the file to use to perform the checksum
+     * @param seed the seed to use for the hasher
+     * @return the checksum hex string of the file's content. XXHash32 is employed as the hash algorithm.
+     */
+    public String checksum(File file, int seed) throws IOException
+    {
+        try (FileInputStream fis = new FileInputStream(file))
+        {
+            try (DigestAlgorithm digestAlgorithm = digestAlgorithmProvider.get(seed))
```
Comment (frankgh) [inline src/main/java/org/apache/cassandra/sidecar/restore/RestoreJobUtil.java]: NIT: 
```suggestion
        try (FileInputStream fis = Files.newInputStream(file.toPath());
               DigestAlgorithm digestAlgorithm = digestAlgorithmProvider.get(seed))
```

**Review by frankgh — APPROVED** [review]
+1 the PR looks good, I have one minor NIT, and suggestions for future improvements

**Pull request closed without merging**